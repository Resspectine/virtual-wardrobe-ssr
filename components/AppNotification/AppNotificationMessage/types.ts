import { Notification } from '@/types/notification';

export interface AppNotificationMessageProps {
  notification: Notification;
  onRemove: (id: number) => void;
}
