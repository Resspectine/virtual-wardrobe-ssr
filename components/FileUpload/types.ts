import { FieldErrors, UseFormRegister, UseFormSetValue } from 'react-hook-form';

import { CreateGarmentValues } from '@/pageComponents/CreateGarment/types';

export interface FileUploadProps {
  register: UseFormRegister<CreateGarmentValues>;
  name: keyof CreateGarmentValues;
  file: File;
  setValue: UseFormSetValue<CreateGarmentValues>;
  disabled?: boolean;
  errors: FieldErrors<CreateGarmentValues>;
}
