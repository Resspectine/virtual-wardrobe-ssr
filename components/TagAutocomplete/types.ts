import { AutocompleteProps } from '@mui/material';

import { Tag } from 'types/tag';

export type CustomTag = { inputValue: string } & Omit<Tag, 'id'>;

export type AnyTag = Tag | CustomTag;

export type TagAutocompleteProps = AutocompleteProps<AnyTag, true, undefined, false>;
