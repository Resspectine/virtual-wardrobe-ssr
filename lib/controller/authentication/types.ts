import { User } from '@/types/user';

export type LoginUser = Pick<User, 'email' | 'password'>;

export type RegisterUser = Pick<User, 'email' | 'password' | 'name'>;
