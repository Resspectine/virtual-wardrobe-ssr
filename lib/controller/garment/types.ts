import { Garment } from '@/types/garment';
import { Tag } from '@/types/tag';

export interface LoadGarmentsQueryParams {
  filter?: string;
  orderBy?: string;
  orderDirection?: string;
}

export type CreateRequestGarment = Omit<Garment, 'id' | 'tags'> & { tags: Omit<Tag, 'id'>[] };

export type UpdateRequestGarment = Omit<Garment, 'tags'> & { tags: (Omit<Tag, 'id'> & { id?: string })[] };
