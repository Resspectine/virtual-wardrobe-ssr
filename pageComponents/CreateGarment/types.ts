export interface CreateGarmentValues {
  title: string;
  description: string;
  price: string;
  image: File[];
}
