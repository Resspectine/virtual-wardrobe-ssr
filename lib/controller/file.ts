export const fileFetch: (file: Blob, url?: string) => Promise<Response> = async (file, url = 'files/upload') => {
  const formData = new FormData();

  formData.append('file', file);

  const request = await fetch(`http://localhost:3030/api/${url}`, {
    method: 'POST',
    body: formData,
    credentials: 'include',
  });

  if (request.status === 401) {
    throw new Error('Unauthorized');
  }

  return request;
};
