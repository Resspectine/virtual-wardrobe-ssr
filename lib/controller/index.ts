export const appFetch: (
  input: RequestInfo,
  init?: (Omit<RequestInit, 'body'> & { body?: Record<string, unknown> }) | undefined,
  returnBlob?: boolean
) => Promise<any> = async (input, init, returnBlob) => {
  const headers = new Headers(init?.headers);

  headers.append('Content-Type', 'application/json');

  const request = await fetch(`http://localhost:3030/api/${input}`, {
    ...init,
    body: JSON.stringify(init?.body || undefined),
    headers,
    credentials: 'include',
  });

  if (request.status === 500) {
    throw new Error('Server error');
  }

  if (!returnBlob) {
    const body = await request.json();

    if (body.statusCode === 401 || body.statusCode === 400) {
      throw new Error(body.message);
    }

    return body;
  }

  return request.blob();
};
