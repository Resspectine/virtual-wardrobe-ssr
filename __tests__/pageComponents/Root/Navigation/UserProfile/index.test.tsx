import { waitFor } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { NextRouter, useRouter } from 'next/router';

import { appFetch } from '@/lib/controller';
import UserProfile from '@/pageComponents/Root/Navigation/UserProfile';
import { placeholderImageUrl } from '@/pageComponents/Root/Navigation/UserProfile/constants';
import { useAppState } from '@/store/appState';
import { StoreUser, useUser } from '@/store/user';
import { render } from '@/testUtils';
import { File } from '@/types/file';

jest.mock('next/router');
jest.mock('@/lib/controller');

const mockedAppFetch = appFetch as unknown as jest.Mock<ReturnType<typeof appFetch>>;
const mockedUseRouter = useRouter as unknown as jest.Mock<ReturnType<typeof useRouter>>;

afterEach(() => {
  mockedAppFetch.mockReset();
  mockedUseRouter.mockReset();
});

afterAll(() => {
  mockedAppFetch.mockRestore();
  mockedUseRouter.mockRestore();
});

describe('UserProfile', () => {
  test('should toggle theme mode, call useRouter push on profile and logout click', async () => {
    const push = jest.fn(() => {});
    mockedUseRouter.mockImplementation(
      () =>
        ({
          push,
        } as unknown as NextRouter)
    );
    mockedAppFetch.mockImplementation(() => Promise.resolve());

    const { result } = renderHook(() => useAppState(state => state.themeMode));

    const { getByRole, getByText, getByAltText } = render(<UserProfile />);

    userEvent.click(getByRole('button'));
    userEvent.click(getByText('Profile'));
    expect(push).toBeCalledTimes(1);

    userEvent.click(getByRole('button'));
    userEvent.click(getByText('Toggle theme'));
    expect(result.current).toBe('red');

    userEvent.click(getByRole('button'));
    userEvent.click(getByText('Logout'));

    await waitFor(() => expect(getByAltText('avatar')).not.toHaveAttribute('src', placeholderImageUrl));

    expect(mockedAppFetch).toBeCalledTimes(1);
  });

  test('should load profile image', async () => {
    mockedAppFetch.mockResolvedValue(new Blob());
    const { result } = renderHook(() => useUser(state => state.setUser));

    result.current({
      avatar: {} as File,
    } as StoreUser);

    const { asFragment, getByAltText } = render(<UserProfile />);

    expect(getByAltText('avatar')).toHaveAttribute('src', placeholderImageUrl);

    expect(mockedAppFetch).toBeCalledTimes(1);

    await waitFor(async () => expect(getByAltText('avatar')).not.toHaveAttribute('src', placeholderImageUrl));

    expect(asFragment()).toMatchSnapshot();
  });
});
