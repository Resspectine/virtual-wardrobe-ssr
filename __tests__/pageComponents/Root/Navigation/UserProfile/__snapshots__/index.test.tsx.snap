// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile should load profile image 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: absolute;
  right: 20px;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: Montserrat,Times New Roman;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 8px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: #1976d2;
  padding: 0;
  min-width: unset;
}

.emotion-1::-moz-focus-inner {
  border-style: none;
}

.emotion-1.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-1 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(25, 118, 210, 0.04);
}

@media (hover: none) {
  .emotion-1:hover {
    background-color: transparent;
  }
}

.emotion-1.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

.emotion-2 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 40px;
  height: 40px;
  font-family: Montserrat,Times New Roman;
  font-size: 1.25rem;
  line-height: 1;
  border-radius: 50%;
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-3 {
  width: 100%;
  height: 100%;
  text-align: center;
  object-fit: cover;
  color: transparent;
  text-indent: 10000px;
}

.emotion-4 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div
    class="MuiBox-root emotion-0"
    data-testid="user-profile-button"
  >
    <button
      class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root emotion-1"
      tabindex="0"
      type="button"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular emotion-2"
      >
        <img
          alt="avatar"
          class="MuiAvatar-img emotion-3"
          src="data:application/octet-stream;base64,"
        />
      </div>
      <span
        class="MuiTouchRipple-root emotion-4"
      />
    </button>
  </div>
</DocumentFragment>
`;
